ethereum/core/src/main/java/tech/pegasys/pantheon/ethereum/storage/keyvalue/KeyValueStorageWorldStateStorage.java

```diff
@@ -29,23 +32,46 @@ public KeyValueStorageWorldStateStorage(final KeyValueStorage keyValueStorage) {
   }
 
   @Override
-  public Optional<BytesValue> getCode(final Hash codeHash) {
-    return keyValueStorage.get(codeHash);
+  public Optional<BytesValue> getCode(final Bytes32 codeHash) {
+    return getCodeValue(codeHash, keyValueStorage::get);
   }
 
   @Override
   public Optional<BytesValue> getAccountStateTrieNode(final Bytes32 nodeHash) {
-    return keyValueStorage.get(nodeHash);
+    return getTrieValue(nodeHash, keyValueStorage::get);
   }
 
   @Override
   public Optional<BytesValue> getAccountStorageTrieNode(final Bytes32 nodeHash) {
-    return keyValueStorage.get(nodeHash);
+    return getTrieValue(nodeHash, keyValueStorage::get);
   }
 
   @Override
-  public Optional<BytesValue> getNodeData(final Hash hash) {
-    return keyValueStorage.get(hash);
+  public Optional<BytesValue> getNodeData(final Bytes32 hash) {
+    return getValue(hash, keyValueStorage::get);
+  }
+
+  private Optional<BytesValue> getTrieValue(
+      final Bytes32 hash, final Function<Bytes32, Optional<BytesValue>> getter) {
+    if (hash.equals(MerklePatriciaTrie.EMPTY_TRIE_NODE_HASH)) {
+      return Optional.of(RLP.NULL);
+    } else {
+      return getter.apply(hash);
+    }
+  }
+
+  private Optional<BytesValue> getCodeValue(
+      final Bytes32 hash, final Function<Bytes32, Optional<BytesValue>> getter) {
+    if (hash.equals(Hash.EMPTY)) {
+      return Optional.of(BytesValue.EMPTY);
+    } else {
+      return getter.apply(hash);
+    }
+  }
+
+  private Optional<BytesValue> getValue(
+      final Bytes32 hash, final Function<Bytes32, Optional<BytesValue>> getter) {
+    return getTrieValue(hash, (h) -> getCodeValue(h, getter));
```

Comment:
I think we've wound up applying the check for empty twice in the `getNodeData` path - once in `WorldStateArchive` and once here.  This is probably the better place, but I actually prefer the more explicit code from `WorldStateArchive` - I know it essentially duplicates some code but explicitly checking for EMPTY and EMPTY_TRIE_NODE_HASH is just a bit more readable than delegating through optionals.

I'm not sure it matters, but it also avoids creating an extra object instances that this lambda version requires (to create a closure over the `getter` param).

Reply:
Fair enough - updated